File: 003/solution.py
# https://projecteuler.net/problem=3
# The prime factors of 13195 are 5, 7, 13 and 29.
# What is the largest prime factor of the number 600851475143 ?
import math

target = 600851475143

# find the max prime factor
max_prime_factor = math.sqrt(target)
max_prime_factor = int(math.floor(max_prime_factor))

primes = []

# build initial 1 based sieve 
lookup = [1] * max_prime_factor

# Build seive
for i in xrange(2, max_prime_factor):
  # if not prime don't mark multiples
  if lookup[i] == 0:
    continue
  # add to prime list
  primes += [i]
  # mark all multiples of the found prime as not prime
  for mark in xrange(i * 2, max_prime_factor, i):
    lookup[mark] = 0

# for each found prime in reverse order check if it is a multiple of the target
for prime in primes[::-1]:
  if target % prime == 0:
    print prime
    break

File: 003/solution.clj
;;; https://projecteuler.net/problem=3
;;; The prime factors of 13195 are 5, 7, 13 and 29.
;;; What is the largest prime factor of the number 600851475143 ?

;; This isn't an ideal solution as each canidate has to be tested for being prime in a slightly better than brute force way

;; Ideally this would use some kind of sieve or at least track known primes for testing both would speed up runtime.

(defn prime? [value]
  "Returns true is value is prime and else otherwise. Test only up to the square root of value"
  (cond
   (= value 2) true ; special case 2
   (< value 2) false 
   :else (let [root (int (Math/sqrt value))] ; sqrt(value) is the max we need to prime for being prime
           (nil? (some #(zero? (mod value %)) (concat [2] (range 3 (inc root) 2)))))))

(println 
  (apply max ; find max of prime factors
         (filter 
           #(zero? (mod 600851475143 %)) ; test for being a factor
           (filter prime? 
                   (concat ; check 2, 3, 5, 7, ... (N+2) ... sqrt(600851475143)
                     [2] 
                     (range 3 (int (Math/sqrt 600851475143)) 2))))))
